clash-prelude/src/Clash/Explicit/Verification.hs

```diff
@@ -0,0 +1,254 @@
+--
+-- where a[n] represents the value of 'a' at cycle 'n' and a[n+m] represents
+-- the value of 'a' at cycle 'n+m'. Cycle n is an arbitrary cycle.
+nextN :: AssertionValue dom a => Word -> a -> Assertion dom
+nextN n = Assertion IsTemporal . CvNext n . assertion . toAssertionValue
+{-# INLINE nextN #-}
+
+-- | Same as @a && next b@ but with a nice syntax. E.g., @a && next b@ could
+-- be written as @a `before` b@. Might be read as "a happens one cycle before b".
+before :: (AssertionValue dom a, AssertionValue dom b) => a -> b -> Assertion dom
+before a0 b0 = Assertion IsTemporal (CvBefore a1 b1)
+ where
+  a1 = assertion (toAssertionValue a0)
+  b1 = assertion (toAssertionValue b0)
+{-# INLINE before #-}
+
+-- | Same as @a `implies` next b@ but with a nice syntax. E.g.,
+-- @a `implies` next b@ could be written as @a `timplies` b@. Might be read
+-- as "a at cycle n implies b at cycle n+1".
+timplies :: (AssertionValue dom a, AssertionValue dom b) => a -> b -> Assertion dom
+timplies a0 b0 = Assertion IsTemporal (CvTemporalImplies 1 a1 b1)
+ where
+  a1 = toTemporal (toAssertionValue a0)
+  b1 = toTemporal (toAssertionValue b0)
+{-# INLINE timplies #-}
+
+-- | Same as 'implies' but strictly temporal.
+timpliesOverlapping
+  :: (AssertionValue dom a, AssertionValue dom b)
+  => a
+  -> b
+  -> Assertion dom
+timpliesOverlapping a0 b0 =
+  Assertion IsTemporal (CvTemporalImplies 0 a1 b1)
+ where
+  a1 = toTemporal (toAssertionValue a0)
+  b1 = toTemporal (toAssertionValue b0)
+{-# INLINE timpliesOverlapping #-}
+
+-- | Specify assertion should _always_ hold
+always :: AssertionValue dom a => a -> Assertion dom
+always = Assertion IsTemporal . CvAlways . assertion . toAssertionValue
+{-# INLINE always #-}
+
+-- | Specify assertion should _never_ hold
+never :: AssertionValue dom a => a -> Assertion dom
+never = Assertion IsTemporal . CvNever . assertion . toAssertionValue
+{-# INLINE never #-}
+
+-- | Check whether given assertion always holds. Results can be collected with
+-- 'check'.
+assert :: AssertionValue dom a => a -> Property dom
+assert = Property . CvAssert . assertion . toAssertionValue
+{-# INLINE assert #-}
+
+-- | Check whether given assertion holds for at least a single cycle. Results
+-- can be collected with 'check'.
+cover :: AssertionValue dom a => a -> Property dom
+cover = Property . CvCover . assertion . toAssertionValue
+{-# INLINE cover #-}
+
+-- | Print property as PSL/SVA in HDL. Clash simulation support not yet
+-- implemented.
+check
+  :: KnownDomain dom
+  => Clock dom
+  -> Reset dom
+  -> Text
+  -- ^ Property name (used in reports and error messages)
+  -> RenderAs
+  -- ^ Assertion language to use in HDL
+  -> Property dom
+  -> Signal dom AssertionResult
+check !_clk !_rst !_propName !_renderAs !_prop =
+  pure (errorX (concat [
+      "Simulation for Clash.Verification not yet implemented. If you need this,"
+    , " create an issue at https://github.com/clash-compiler/clash-lang/issues." ]))
```

Comment:
Is there a reason this isn't done with the interpolator? I know it doesn't really need it, but I would like diagnostic reporting to gradually become more uniform so it's easier to improve it later

Reply:
Yes! I left a comment on your earlier comment, but GitHub makes it _really_ easy to miss:

> The new quasiquoter is in clash-lib :'(. But I've added the msg

Reply:
Ah yes, of course. Nevermind then.